File: src/crimson/weapon_runtime/fire_recipes.py
from __future__ import annotations

import msgspec

from ..effects import ParticleStyleId
from ..projectiles.types import ProjectileTemplateId, SecondaryProjectileTypeId
from ..weapons import WeaponId, projectile_type_id_for_weapon_id


class NoJitter(msgspec.Struct, frozen=True, tag=True):
    pass


class ModuloCenteredJitter(msgspec.Struct, frozen=True, tag=True):
    modulo: int
    center: int
    step: float


class MaskCenteredJitter(msgspec.Struct, frozen=True, tag=True):
    mask: int
    center: int
    step: float


type PelletJitterRule = NoJitter | ModuloCenteredJitter | MaskCenteredJitter


class NoSpeedScale(msgspec.Struct, frozen=True, tag=True):
    pass


class ModuloSpeedScale(msgspec.Struct, frozen=True, tag=True):
    base: float
    modulo: int
    step: float


type SpeedScaleRule = NoSpeedScale | ModuloSpeedScale


class NoTargetHint(msgspec.Struct, frozen=True, tag=True):
    pass


class UseAimTargetHint(msgspec.Struct, frozen=True, tag=True):
    pass


type SecondaryTargetingPolicy = NoTargetHint | UseAimTargetHint


class PrimaryPelletsMode(msgspec.Struct, frozen=True, tag=True):
    type_id: ProjectileTemplateId | None = None
    count: int | None = None
    jitter: PelletJitterRule = msgspec.field(default_factory=NoJitter)
    speed_scale: SpeedScaleRule = msgspec.field(default_factory=NoSpeedScale)


class SecondaryShotMode(msgspec.Struct, frozen=True, tag=True):
    type_id: SecondaryProjectileTypeId
    targeting: SecondaryTargetingPolicy = msgspec.field(default_factory=NoTargetHint)


class ParticleStreamMode(msgspec.Struct, frozen=True, tag=True):
    style: ParticleStyleId | None = None
    slow: bool = False


class MultiPlasmaFanMode(msgspec.Struct, frozen=True, tag=True):
    pass


class SwarmerDumpMode(msgspec.Struct, frozen=True, tag=True):
    pass


type FireMode = PrimaryPelletsMode | SecondaryShotMode | ParticleStreamMode | MultiPlasmaFanMode | SwarmerDumpMode


class FireRecipe(msgspec.Struct, frozen=True):
    mode: FireMode
    ammo_cost: float = 1.0


def pellet_jitter_step_for_weapon(weapon_id: WeaponId) -> float:
    weapon_id = WeaponId(weapon_id)
    if weapon_id == WeaponId.SHOTGUN:
        return 0.0013
    if weapon_id == WeaponId.SAWED_OFF_SHOTGUN:
        return 0.004
    if weapon_id == WeaponId.JACKHAMMER:
        return 0.0013
    return 0.0015


_SHOTGUN_SPEED_RANDOMIZE_WEAPONS: frozenset[WeaponId] = frozenset(
    {
        WeaponId.SHOTGUN,
        WeaponId.SAWED_OFF_SHOTGUN,
        WeaponId.JACKHAMMER,
    },
)


_DEFAULT_SPREAD_JITTER = ModuloCenteredJitter(modulo=200, center=100, step=0.0015)
_DEFAULT_SPEED_SCALE = ModuloSpeedScale(base=1.0, modulo=100, step=0.01)
_GAUSS_ION_SPEED_SCALE = ModuloSpeedScale(base=1.4, modulo=0x50, step=0.01)


FIRE_RECIPE_BY_WEAPON: dict[WeaponId, FireRecipe] = {
    WeaponId.ROCKET_LAUNCHER: FireRecipe(mode=SecondaryShotMode(type_id=SecondaryProjectileTypeId.ROCKET)),
    WeaponId.SEEKER_ROCKETS: FireRecipe(
        mode=SecondaryShotMode(
            type_id=SecondaryProjectileTypeId.HOMING_ROCKET,
            targeting=UseAimTargetHint(),
        ),
    ),
    WeaponId.ROCKET_MINIGUN: FireRecipe(mode=SecondaryShotMode(type_id=SecondaryProjectileTypeId.ROCKET_MINIGUN)),
    WeaponId.FLAMETHROWER: FireRecipe(
        mode=ParticleStreamMode(style=None, slow=False),
        ammo_cost=0.1,
    ),
    WeaponId.BLOW_TORCH: FireRecipe(
        mode=ParticleStreamMode(style=ParticleStyleId.BLOW_TORCH, slow=False),
        ammo_cost=0.05,
    ),
    WeaponId.HR_FLAMER: FireRecipe(
        mode=ParticleStreamMode(style=ParticleStyleId.HR_FLAMER, slow=False),
        ammo_cost=0.1,
    ),
    WeaponId.BUBBLEGUN: FireRecipe(
        mode=ParticleStreamMode(style=None, slow=True),
        ammo_cost=0.15,
    ),
    WeaponId.MULTI_PLASMA: FireRecipe(mode=MultiPlasmaFanMode()),
    WeaponId.MINI_ROCKET_SWARMERS: FireRecipe(mode=SwarmerDumpMode()),
    WeaponId.PLASMA_SHOTGUN: FireRecipe(
        mode=PrimaryPelletsMode(
            type_id=ProjectileTemplateId.PLASMA_MINIGUN,
            count=14,
            jitter=MaskCenteredJitter(mask=0xFF, center=0x80, step=0.002),
            speed_scale=_DEFAULT_SPEED_SCALE,
        ),
    ),
    WeaponId.GAUSS_SHOTGUN: FireRecipe(
        mode=PrimaryPelletsMode(
            type_id=ProjectileTemplateId.GAUSS_GUN,
            count=6,
            jitter=ModuloCenteredJitter(modulo=200, center=100, step=0.002),
            speed_scale=_GAUSS_ION_SPEED_SCALE,
        ),
    ),
    WeaponId.ION_SHOTGUN: FireRecipe(
        mode=PrimaryPelletsMode(
            type_id=ProjectileTemplateId.ION_MINIGUN,
            count=8,
            jitter=ModuloCenteredJitter(modulo=200, center=100, step=0.0026),
            speed_scale=_GAUSS_ION_SPEED_SCALE,
        ),
    ),
}


def resolve_fire_recipe(
    weapon_id: WeaponId,
    *,
    pellet_count: int,
    fire_bullets_active: bool,
) -> FireRecipe:
    if fire_bullets_active:
        return FireRecipe(
            mode=PrimaryPelletsMode(
                type_id=ProjectileTemplateId.FIRE_BULLETS,
                count=max(0, int(pellet_count)),
                jitter=_DEFAULT_SPREAD_JITTER,
                speed_scale=NoSpeedScale(),
            ),
        )

    recipe = FIRE_RECIPE_BY_WEAPON.get(WeaponId(weapon_id))
    if recipe is not None:
        return recipe

    pellets = max(1, int(pellet_count))
    jitter: PelletJitterRule = NoJitter()
    if pellets > 1:
        jitter = ModuloCenteredJitter(
            modulo=200,
            center=100,
            step=pellet_jitter_step_for_weapon(weapon_id),
        )

    speed_scale: SpeedScaleRule = NoSpeedScale()
    if pellets > 1 and WeaponId(weapon_id) in _SHOTGUN_SPEED_RANDOMIZE_WEAPONS:
        speed_scale = _DEFAULT_SPEED_SCALE

    return FireRecipe(
        mode=PrimaryPelletsMode(
            type_id=projectile_type_id_for_weapon_id(WeaponId(weapon_id)),
            count=pellets,
            jitter=jitter,
            speed_scale=speed_scale,
        ),
    )


__all__ = [
    "FIRE_RECIPE_BY_WEAPON",
    "FireMode",
    "FireRecipe",
    "MaskCenteredJitter",
    "ModuloCenteredJitter",
    "ModuloSpeedScale",
    "MultiPlasmaFanMode",
    "NoJitter",
    "NoSpeedScale",
    "NoTargetHint",
    "ParticleStreamMode",
    "PelletJitterRule",
    "PrimaryPelletsMode",
    "SecondaryShotMode",
    "SecondaryTargetingPolicy",
    "SpeedScaleRule",
    "SwarmerDumpMode",
    "UseAimTargetHint",
    "pellet_jitter_step_for_weapon",
    "resolve_fire_recipe",
]

File: crimson-zig/src/runtime/fire_recipes.zig
const game_ids = @import("../game_ids.zig");

const particles_mod = @import("particles.zig");
const secondary_projectiles_mod = @import("secondary_projectiles.zig");
const weapon_data = @import("weapon_data.zig");

pub const WeaponId = game_ids.WeaponId;
pub const ProjectileTypeId = game_ids.ProjectileTypeId;

pub const PelletJitterRule = union(enum) {
    none,
    modulo_centered: struct {
        modulo: u32,
        center: i32,
        step: f32,
    },
    mask_centered: struct {
        mask: u32,
        center: i32,
        step: f32,
    },
};

pub const SpeedScaleRule = union(enum) {
    none,
    modulo: struct {
        base: f32,
        modulo: u32,
        step: f32,
    },
};

pub const SecondaryTargetingPolicy = enum {
    none,
    use_aim_target_hint,
};

pub const PrimaryPelletsMode = struct {
    type_id: ProjectileTypeId,
    count: i32,
    jitter: PelletJitterRule,
    speed_scale: SpeedScaleRule,
};

pub const SecondaryShotMode = struct {
    type_id: secondary_projectiles_mod.SecondaryProjectileTypeId,
    targeting: SecondaryTargetingPolicy = .none,
};

pub const ParticleStreamMode = struct {
    style: ?particles_mod.ParticleStyleId = null,
    slow: bool = false,
};

pub const FireMode = union(enum) {
    primary_pellets: PrimaryPelletsMode,
    secondary_shot: SecondaryShotMode,
    particle_stream: ParticleStreamMode,
    multi_plasma_fan: void,
    swarmer_dump: void,
};

pub const FireRecipe = struct {
    mode: FireMode,
    ammo_cost: f32 = 1.0,
};

const default_spread_jitter: PelletJitterRule = .{
    .modulo_centered = .{
        .modulo = 200,
        .center = 100,
        .step = 0.0015,
    },
};

const default_speed_scale: SpeedScaleRule = .{
    .modulo = .{
        .base = 1.0,
        .modulo = 100,
        .step = 0.01,
    },
};

const gauss_ion_speed_scale: SpeedScaleRule = .{
    .modulo = .{
        .base = 1.4,
        .modulo = 0x50,
        .step = 0.01,
    },
};

pub fn pelletJitterStepForWeapon(weapon_id: WeaponId) f32 {
    return switch (weapon_id) {
        .shotgun, .jackhammer => 0.0013,
        .sawed_off_shotgun => 0.004,
        else => 0.0015,
    };
}

pub fn resolveFireRecipe(
    weapon_id: WeaponId,
    pellet_count: i32,
    fire_bullets_active: bool,
) FireRecipe {
    if (fire_bullets_active) {
        return .{
            .mode = .{
                .primary_pellets = .{
                    .type_id = .fire_bullets,
                    .count = @max(0, pellet_count),
                    .jitter = default_spread_jitter,
                    .speed_scale = .none,
                },
            },
        };
    }

    return switch (weapon_id) {
        .rocket_launcher => .{
            .mode = .{
                .secondary_shot = .{
                    .type_id = .rocket,
                },
            },
        },
        .seeker_rockets => .{
            .mode = .{
                .secondary_shot = .{
                    .type_id = .homing_rocket,
                    .targeting = .use_aim_target_hint,
                },
            },
        },
        .rocket_minigun => .{
            .mode = .{
                .secondary_shot = .{
                    .type_id = .rocket_minigun,
                },
            },
        },
        .flamethrower => .{
            .mode = .{
                .particle_stream = .{},
            },
            .ammo_cost = 0.1,
        },
        .blow_torch => .{
            .mode = .{
                .particle_stream = .{
                    .style = .blow_torch,
                },
            },
            .ammo_cost = 0.05,
        },
        .hr_flamer => .{
            .mode = .{
                .particle_stream = .{
                    .style = .hr_flamer,
                },
            },
            .ammo_cost = 0.1,
        },
        .bubblegun => .{
            .mode = .{
                .particle_stream = .{
                    .slow = true,
                },
            },
            .ammo_cost = 0.15,
        },
        .multi_plasma => .{
            .mode = .{ .multi_plasma_fan = {} },
        },
        .mini_rocket_swarmers => .{
            .mode = .{ .swarmer_dump = {} },
        },
        .plasma_shotgun => .{
            .mode = .{
                .primary_pellets = .{
                    .type_id = .plasma_minigun,
                    .count = 14,
                    .jitter = .{
                        .mask_centered = .{
                            .mask = 0xff,
                            .center = 0x80,
                            .step = 0.002,
                        },
                    },
                    .speed_scale = default_speed_scale,
                },
            },
        },
        .gauss_shotgun => .{
            .mode = .{
                .primary_pellets = .{
                    .type_id = .gauss_gun,
                    .count = 6,
                    .jitter = .{
                        .modulo_centered = .{
                            .modulo = 200,
                            .center = 100,
                            .step = 0.002,
                        },
                    },
                    .speed_scale = gauss_ion_speed_scale,
                },
            },
        },
        .ion_shotgun => .{
            .mode = .{
                .primary_pellets = .{
                    .type_id = .ion_minigun,
                    .count = 8,
                    .jitter = .{
                        .modulo_centered = .{
                            .modulo = 200,
                            .center = 100,
                            .step = 0.0026,
                        },
                    },
                    .speed_scale = gauss_ion_speed_scale,
                },
            },
        },
        else => blk: {
            const pellets = @max(1, pellet_count);
            const jitter: PelletJitterRule = if (pellets > 1)
                .{
                    .modulo_centered = .{
                        .modulo = 200,
                        .center = 100,
                        .step = pelletJitterStepForWeapon(weapon_id),
                    },
                }
            else
                .none;

            const speed_scale: SpeedScaleRule = if (pellets > 1 and
                (weapon_id == .shotgun or weapon_id == .sawed_off_shotgun or weapon_id == .jackhammer))
                default_speed_scale
            else
                .none;

            break :blk .{
                .mode = .{
                    .primary_pellets = .{
                        .type_id = weapon_data.projectileTypeIdFromWeaponId(weapon_id) orelse unreachable,
                        .count = pellets,
                        .jitter = jitter,
                        .speed_scale = speed_scale,
                    },
                },
            };
        },
    };
}
